### PR TITLE
Critical Commanding Support For Enterprise

### PR DESCRIPTION
**Pull request by ryanmelt**

### CODE REVIEW

File: openc3-cosmos-cmd-tlm-api/app/models/messages_thread.rb [hunk 1]
```diff
@@ -43,10 +43,10 @@ def initialize(
     @redis_offset = nil # Redis offset to transition from files
     @scope = scope
     @thread_mode = :SETUP
-    @topics = ["#{scope}__openc3_log_messages"]
+    @topics = ["#{scope}__openc3_log_messages", "#{scope}__openc3_ephemeral_messages"]
 
     offsets = nil
-    offsets = [start_offset] if start_offset
+    offsets = [start_offset, "$"] if start_offset
```
Comment (jmthomas) [inline openc3-cosmos-cmd-tlm-api/app/models/messages_thread.rb:49]: What does the "$" do? comment?

File: openc3/lib/openc3/models/scope_model.rb [hunk 2]
```diff
@@ -252,6 +276,21 @@ def deploy_scopecleanup_microservice(gem_path, variables, parent)
       Logger.info "Configured microservice #{microservice_name}"
     end
 
+    def deploy_critical_cmd_microservice(gem_path, variables, parent)
+      microservice_name = "#{@scope}__CRITICALCMD__#{@scope}"
```
Comment (jmthomas) [inline openc3/lib/openc3/models/scope_model.rb:280]: Need a corresponding `model.destroy` in `undeploy`

File: openc3/lib/openc3/script/commands.rb [hunk 3]
```diff
@@ -120,17 +120,25 @@ def _cmd(cmd, cmd_no_hazardous, *args, timeout: nil, log_message: nil, validate:
...
+            target_name, cmd_name, cmd_params = $api_server.method_missing(cmd, *args, timeout: timeout, log_message: log_message, validate: validate, scope: scope, token: token)
+            if log_message.nil? or log_message
+              _log_cmd(target_name, cmd_name, cmd_params, raw, no_range, no_hazardous)
+            end
+          rescue HazardousError => e
+            # This opens a prompt at which point they can cancel and stop the script
+            # or say Yes and send the command. Thus we don't care about the return value.
+            prompt_for_hazardous(e.target_name, e.cmd_name, e.hazardous_description)
+            target_name, cmd_name, cmd_params = $api_server.method_missing(cmd_no_hazardous, *args, timeout: timeout, log_message: log_message, validate: validate, scope: scope, token: token)
+            if log_message.nil? or log_message
```
Comment (jmthomas) [inline openc3/lib/openc3/script/commands.rb:133]: This is a weird check but I think it means log_message must be explicitly `false` for us to NOT call `_log_cmd`

**Review by jmthomas — CHANGES_REQUESTED** [review]
Fix spelling, ruby tests, add delete microservice, and python code

File: openc3-cosmos-init/plugins/packages/openc3-tool-common/src/components/CriticalCmdDialog.vue [hunk 4]
```diff
@@ -0,0 +1,263 @@
...
+              label="Password"
+              data-test="password"
+            />
+          </v-row>
+          <v-row>
+            <v-spacer />
+            <v-btn
+              type="submit"
+              @click.prevent="reject"
+              class="mx-2"
```
Comment (jmthomas) [inline openc3-cosmos-init/plugins/packages/openc3-tool-common/src/components/CriticalCmdDialog.vue:54]: Add `outlined` to match existing dialogs

Comment (ryanmelt) [reply openc3-cosmos-init/plugins/packages/openc3-tool-common/src/components/CriticalCmdDialog.vue:54]: I had this originally, but it didn't respect disabled (looked exactly the same), so I dropped it.

**Comment by jmthomas** [comment]
I noticed when sending a command in Command Sender you initially get the status as:
`Status: Error sending INST CLEAR due to CriticalCmdError: CriticalCmdError`

Even after approving the command the status doesn't change. It should say the command was sent. Not sure if you have to listen to the topic to know when this approval happens or what but we need the status to be accurate.

**Comment by jmthomas** [comment]
Updated confirmation dialog that handles errors:
Screenshot 2024-10-10 at 9 26 23 PM
Screenshot 2024-10-10 at 9 26 39 PM
Screenshot 2024-10-10 at 9 26 51 PM
Screenshot 2024-10-10 at 9 27 02 PM

File: openc3-cosmos-cmd-tlm-api/app/models/messages_thread.rb (hunk 1 above)
Comment (ryanmelt) [reply openc3-cosmos-cmd-tlm-api/app/models/messages_thread.rb:49]: Added comment.

File: openc3/lib/openc3/models/scope_model.rb (hunk 2 above)
Comment (ryanmelt) [reply openc3/lib/openc3/models/scope_model.rb:280]: Good catch.

**Pull request merged**